build/Jamfile

```diff
@@ -36,6 +36,8 @@ py3-version = [ py-version 3 ] ;
 
 project boost/python
   : source-location ../src
+  : requirements <toolset>gcc <target-os>windows:<cxxflags>-D_hypot=hypot
+  : requirements <toolset>gcc <target-os>windows <link>shared:<cxxflags>-DMS_WIN64
```

Comment:
I don't think this is doing what you expect. As far as I understand Boost.Build syntax, the last line above defines as requireents:
* toolset=gcc
* target-os=windows
* cxxflags=-DMS_WIN64, whenever link=shared

It looks like what you intend to do is to define -DMS_WIN64 whenever all three features (toolset, target-os, and link) have the given values, yes ? 
Please ask on the boost or boost.build mailing list what the right syntax for this is.
(I'm also not sure how this would fix the `"gcc is not a known value..." error. That's another thing to ask on boost.build ML).
Sorry I can't help more with this. Boost.Build is complicated and badly documented, which is one reason why I'm trying to move away from it...